core/state/pruner/pruner.go

```diff
@@ -401,54 +404,121 @@ func dumpRawTrieDescendants(db ethdb.Database, root common.Hash, output *stateBl
+						if err != nil {
+							return
+						}
+						storageIt, err := storageTr.NodeIterator(nil)
+						if err != nil {
+							return
+						}
+						var processedNodes uint64
+						for storageIt.Next(true) {
+							storageTrieHash := storageIt.Hash()
+							if storageTrieHash != (common.Hash{}) {
+								// The inner bloomfilter library has a mutex so concurrency is fine here
+								err = output.Put(storageTrieHash.Bytes(), nil)
+								if err != nil {
+									return
+								}
+							}
+							processedNodes++
+							if time.Since(threadLastLog) > 5*time.Minute {
+								elapsedTotal := time.Since(startedAt)
+								elapsedThread := time.Since(threadStartedAt)
+								log.Info("traversing trie database - traversing storage trie taking long", "key", key, "elapsedTotal", elapsedTotal, "elapsedThread", elapsedThread, "processedNodes", processedNodes, "threadsRunning", threadsRunning.Load())
+								threadLastLog = time.Now()
 							}
 						}
-						processedNodes++
-						if time.Since(threadLastLog) > 5*time.Minute {
-							elapsedTotal := time.Since(startedAt)
-							elapsedThread := time.Since(threadStartedAt)
-							log.Info("traversing trie database - traversing storage trie taking long", "key", key, "elapsedTotal", elapsedTotal, "elapsedThread", elapsedThread, "processedNodes", processedNodes, "threadsRunning", threadsRunning.Load())
-							threadLastLog = time.Now()
+						err = storageIt.Error()
+						if err != nil {
+							return
 						}
+					}()
+				} else {
+					for i := int64(1); i <= 32; i++ {
+						err = <-results
+						if err != nil {
+							return err
+						}
+						go func(iteration int64) {
+							threadsRunning.Add(1)
+							defer threadsRunning.Add(-1)
+							var processedNodes uint64
+							threadStartedAt := time.Now()
+							threadLastLog := time.Now()
+							var err error
+							defer func() {
+								results <- err
+							}()
+							// note: we are passing data.Root as stateRoot here, to skip the check for stateRoot existence in trie.newTrieReader,
+							// we already check that when opening state trie and reading the account node
+							// Need to create a new trie for each iteration, to avoid race conditions.
+							trieID := trie.StorageTrieID(data.Root, key, data.Root)
+							var storageTr *trie.StateTrie
+							storageTr, err = trie.NewStateTrie(trieID, sdb.TrieDB())
+							if err != nil {
+								return
+							}
+							var startIt trie.NodeIterator
+							startIt, err = storageTr.NodeIterator(big.NewInt((iteration - 1) << 3).Bytes())
+							if err != nil {
+								return
+							}
+							// Traverse the storage trie, and stop if we reach the end of the trie or the end of the current part
+							var startItPath, endItPath []byte
+
+							key := keybytesToHex(big.NewInt((iteration) << 3).Bytes())
+							key = key[:len(key)-1]
+							for startIt.Next(true) {
+								if iteration != 32 && bytes.Compare(startIt.Path(), key) >= 0 {
```

Comment:
shouldn't the second condition be `bytes.Compare(startIt.Path(), key) > 0`?

the NodeIterator starts at the key after the given start key, am I right that with >= we might skip the trie node with `path == key` ?

https://github.com/OffchainLabs/go-ethereum/blob/8070936269ea11801bb91e90dfcf66685cba9fcc/trie/secure_trie.go#L253-L257

Reply:
Thanks for catching this, updated!